docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+   will later detect the mismatch, and correct it by performing the
+   complementary operation of the relocation (a remove or a create).
+
+The first piece is already present: each range has a zone configuration
+which determines the target replication state for the range.
+
+The second piece, the core mechanic, will be performed by the replicate
+queue. This queue is already used to add replicas to ranges which are
+under-replicated; it can be enhanced to remove replicas from over-replicated
+ranges, thus satisfying the basic requirements of the core mechanic. Detailed
+design for these changes are in a [separate RFC (#2153)](https://github.com/cockroachdb/cockroach/pull/2153).
+
+The third piece simply informs the design of systems performing relocations:
+most immediately, these will be the rebalance system and the repair system.
+These systems, when identifying a relocation opportunity, will simply perform
+the first operation of the relocation, and the replicate queue will perform the
+complementary operation. 
+
+RFCs for these operations are forthcoming.
+
+The final complication is how to ensure that the replicate queue promptly
+identifies ranges outside of their ideal target state. As a queue it will be
+attached to the replica scanner, but we will also want to enqueue a replica
+immediately whenever we knowingly perturb the replication state. Thus,
+components initating a relocation (e.g. rebalance or repair) should immediately
+enqueue their target replicas after changing the replication state. 
+
+# Drawbacks
+
+### Specific Move Operations
+A stateless model for relocation precludes the ability to request specific
+relocations; only the first operation can be made with specificity.
+
+For example, the verb "Move replica from X to Y" cannot be expressed with
+certainty; instead, only "Move replica to X from (some store)" or "Move replica
+to Y from (some store)" can be expressed. The replicate queue will be responsible
+for selecting an appropriate store for the complementary operation.
+
+It is assumed that this level of specificity is simply not needed for any
+relocation operations; there is no currently apparent use case where a move
+between a specific pair of stores is needed.
+
+Even if this was necessary, it might be possible to express by manipulating the
+factors behind the individual stateless decisions.
+
+### Thrashing of complementary operation
+Because there is no relocation state, the possibilty of "thrashing" is introduced. For
+example: 
+
+1. Rebalance operation adds a new replica to the range.
+2. The replicate queue picks up the range and detects the need to remove a
+   replica; however, it decides to remove the replica that was just added.
+
+This is possible if the rebalance's criteria for new replica selection are
+sufficiently different from the replicate queue's selection criteria for
+removing a replica.
+
+To reduce this risk, there must be sufficient agreement in the criteria between
```

Comment:
What if there is an inherent asymmetry in the criteria? If we remove a node whose raft position falls behind (but otherwise appears healthy), it will still be a candidate to be its own replacement, since we there are no raft positions to consider in new replicas. Maybe that's not a very likely problem, but expecting symmetry in the scores could be a bit risky.